core/src/main/java/org/infinispan/cache/impl/CacheImpl.java

```diff
@@ -557,6 +558,14 @@ private V removeInternal(Object key, EnumSet<Flag> explicitFlags, InvocationCont
       return (V) executeCommandAndCommitIfNeeded(ctx, command);
    }
 
+   @Override
+   public void removeExpired(K key, V value, Long lifespan) {
+      InvocationContext ctx = getInvocationContextWithImplicitTransaction(false, null, 1);
+      RemoveExpiredCommand command = commandsFactory.buildRemoveExpiredCommand(key, value, lifespan);
+      ctx.setLockOwner(command.getKeyLockOwner());
```

Comment:
I'm not sure about this... The key locks are no longer reentrant, so unlocking the key for the `RemoveExpiredCommand` will unlock it for the transaction/operation that triggered it as well.